control/statefbk.py

```diff
@@ -326,11 +341,61 @@ def lqe(A, G, C, QN, RN, NN=None):
     #    NN = np.zeros(QN.size(0),RN.size(1))
     # NG = G @ NN
 
-    # LT, P, E = lqr(A.T, C.T, G @ QN @ G.T, RN)
-    # P, E, LT = care(A.T, C.T, G @ QN @ G.T, RN)
-    A, G, C = np.array(A, ndmin=2), np.array(G, ndmin=2), np.array(C, ndmin=2)
-    QN, RN = np.array(QN, ndmin=2), np.array(RN, ndmin=2)
-    P, E, LT = care(A.T, C.T, np.dot(np.dot(G, QN), G.T), RN)
+    #
+    # Process the arguments and figure out what inputs we received
+    #
+
+    # Get the system description
+    if (len(args) < 3):
+        raise ControlArgument("not enough input arguments")
+
+    try:
+        sys = args[0]           # Treat the first argument as a system
+        if isinstance(sys, LTI):
+            # Convert LTI system to state space
+            sys = _convert_to_statespace(sys)
+
+        # Extract A, G (assume disturbances come through input), and C
+        A = np.array(sys.A, ndmin=2, dtype=float)
+        G = np.array(sys.B, ndmin=2, dtype=float)
+        C = np.array(sys.C, ndmin=2, dtype=float)
+        index = 1
+
+    except AttributeError:
+        # Arguments should be A and B matrices
+        A = np.array(args[0], ndmin=2, dtype=float)
+        G = np.array(args[1], ndmin=2, dtype=float)
+        C = np.array(args[2], ndmin=2, dtype=float)
+        index = 3
+
+    # Get the weighting matrices (converting to matrices, if needed)
+    Q = np.array(args[index], ndmin=2, dtype=float)
+    R = np.array(args[index+1], ndmin=2, dtype=float)
+
+    # Get the cross-covariance matrix, if given
+    if (len(args) > index + 2):
+        N = np.array(args[index+2], ndmin=2, dtype=float)
+        raise ControlNotImplemented("cross-covariance not implemented")
+
+    else:
+        N = np.zeros((Q.shape[0], R.shape[1]))
+
+    # Check dimensions for consistency
+    nstates = A.shape[0]
+    ninputs = G.shape[1]
+    noutputs = C.shape[0]
+    if (A.shape[0] != nstates or A.shape[1] != nstates or
+        G.shape[0] != nstates or C.shape[1] != nstates):
+        raise ControlDimension("inconsistent system dimensions")
+
+    elif (Q.shape[0] != ninputs or Q.shape[1] != ninputs or
+          R.shape[0] != noutputs or R.shape[1] != noutputs or
+          N.shape[0] != ninputs or N.shape[1] != noutputs):
+        raise ControlDimension("incorrect covariance matrix dimensions")
+
+    # LT, P, E = lqr(A.T, C.T, G @ Q @ G.T, R)
```

Comment:
Is this correct, or should it rather be
```suggestion
    # L, P, E = lqe(...)
```

?

Reply:
This was a comment (mostly left for myself) that it is also possible to get the LQE gains by using the lqr function as shown in this comment. LT means it's the transpose of the L you want.  Probably best just to remove the comment to avoid confusion.